biorbd_optim/constraints.py

```diff
@@ -221,22 +229,52 @@ def __contact_force_inequality(ocp, nlp, X, U, policy, _type):
             ["CS"],
         ).expand()
 
-        if not isinstance(policy[0], (tuple, list)):
-            policy = [policy]
+        X, U = horzsplit(X, 1), horzsplit(U, 1)
+        for i in range(len(U)):
+            ocp.g = vertcat(ocp.g, CS_func(X[i], U[i])[idx])
+            if _type == "GREATER_THAN":
+                ocp.g_bounds.min.append(boundary)
+                ocp.g_bounds.max.append(inf)
+            elif _type == "LESSER_THAN":
+                ocp.g_bounds.min.append(-inf)
+                ocp.g_bounds.max.append(boundary)
 
+    @staticmethod
+    def __non_slipping(ocp, nlp, X, U, normal_component_idx, tangential_component_idx, static_friction_coefficient):
+        """
+        :param coeff: It is the coefficient of static friction.
+        """
+        CS_func = Function(
+            "Contact_force_inequality",
+            [ocp.symbolic_states, ocp.symbolic_controls],
+            [Dynamics.forces_from_forward_dynamics_with_contact(ocp.symbolic_states, ocp.symbolic_controls, nlp)],
+            ["x", "u"],
+            ["CS"],
+        ).expand()
+
+        mu = static_friction_coefficient
         X, U = horzsplit(X, 1), horzsplit(U, 1)
         for i in range(len(U)):
-            contact_forces = CS_func(X[i], U[i])
-            contact_forces = contact_forces[: nlp["model"].nbContacts()]
-
-            for elem in policy:
-                ocp.g = vertcat(ocp.g, contact_forces[elem[0]])
-                if _type == "GREATER_THAN":
-                    ocp.g_bounds.min.append(elem[1])
-                    ocp.g_bounds.max.append(inf)
-                elif _type == "LESSER_THAN":
-                    ocp.g_bounds.min.append(-inf)
-                    ocp.g_bounds.max.append(elem[1])
+            normal_contact_force = tangential_contact_force = 0
+            for idx in normal_component_idx:
+                normal_contact_force += CS_func(X[i], U[i])[idx]
+            for idx in tangential_component_idx:
+                normal_contact_force += CS_func(X[i], U[i])[idx]
```

Comment:
add `forces = CS_func(X[i], U[i])` before the for loops, and use forces instead. I am not sure if it is better, but I feel that CS_func won't be reevaluated

Reply:
or 
`sum(CS_func(X[i], U[i])[normal_component_idx])`